internal/pipeline/executor.go

```diff
@@ -36,159 +38,129 @@ type Executor struct {
+		return err
 	}
-
 	// initialize hydration context
 	hctx := &hydrationContext{
-		root:            p,
-		hydrationStates: map[string]hydrationState{},
-		pkgs:            map[string]*pkg{},
+		root: rootPkg,
+		pkgs: map[string]*pkgNode{},
 	}
-
-	resources, err := hydrate(p, hctx)
+	resources, err := hydrate(rootPkg, hctx)
 	if err != nil {
-		return fmt.Errorf("failed to hydrate the pkg %s %w", e.PkgPath, err)
+		return fmt.Errorf("failed to hydrate the pkg %q %w", rootPkg.Path(), err)
 	}
-	// TODO(droot): support different sink modes
-	// for now do in-place
-	klog.Infof("hydrated reources: %d", len(resources))
 
+	pkgWriter := &kio.LocalPackageWriter{PackagePath: rootPkg.Path()}
+	err = pkgWriter.Write(resources)
+	if err != nil {
+		klog.Errorf("failed to write pkg: %v", err)
+		return fmt.Errorf("failed to write pkg: %w", err)
+	}
 	return nil
 }
 
+// helper function to log debug information about a resource.
+func debugResource(r *yaml.RNode) {
+	meta, _ := r.GetMeta()
+	klog.Infof("resource %s annotations: %v", meta.Name, meta.Annotations)
+}
+
 //
 // hydrationContext contains bits to track state of a package hydration.
 // This is sort of global state that is available to hydration step at
 // each pkg along the hydration walk.
 type hydrationContext struct {
 	// root points to the root of hydration graph where we bagan our hydration journey
-	root *pkg
-
-	// TODO (droot): wire sink-mode here
+	root *pkgNode
 
-	// pkgs refers to the discovered pkgs where
-	// It's a map where key refers to package's absolute path and value
-	// refers to the pkg.
-	pkgs map[string]*pkg
-
-	// hydrationStates for each package (dry, hydrating or wet) in the hydration DAG
-	hydrationStates map[string]hydrationState
+	// pkgs refers to the packages undergoing hydration.
+	pkgs map[string]*pkgNode
 }
 
 //
-// pkg represents a kpt package (node) in the hydration DAG.
+// pkgNode represents a package being hydrated. Think of it as a node in the hydration DAG.
 //
-// TODO(droot): kpt will certainly have pkg abstraction, may be embed that here to
-// avoid duplication.
-type pkg struct {
-	// absolute path on the filesystem
-	path string
+type pkgNode struct {
+	pkg *pkg.Pkg
 
-	// pipeline contained in the pkg
-	pipeline *kptfilev1alpha2.Pipeline
+	// state indicates if the pkg is being hydrated or done.
+	state hydrationState
 
-	// resources post hydration
+	// KRM resources that we have gathered post hydration for this package.
+	// These inludes resources at this pkg as well all it's children.
 	resources []*yaml.RNode
 }
 
-// TODO(droot): this will be replaced when pkg abstraction PR is merged
-func newPkg(path string) (*pkg, error) {
-	pkgPath, err := filepath.Abs(path)
+// newPkgNode returns a pkgNode instance given a path.
+func newPkgNode(path string) (*pkgNode, error) {
+	p, err := pkg.New(path)
 	if err != nil {
-		return nil, fmt.Errorf("failed to find abs path %s %w", path, err)
+		return nil, fmt.Errorf("failed to read pkg %w", err)
```

Comment:
general comment: choose the failure error messages shown to the user carefully. It's an important part of the UX. In this case, `pkg` is an internal variable name.

Reply:
I agree. I have improved error messages to make it useful for the user for now, but I will scrub them again once we have the finalized the command UX and terminology.

I realize that errors are serving two users (implementor and the end-user), so striking the balance sometimes is getting difficult. I found this post from Rob Pike https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html very useful. I want to explore if we can apply those lessons to kpt error handling.